app/ece.js

```diff
@@ -0,0 +1,292 @@
+      controller.enqueue(this.createHeader());
+    } else if (this.mode !== MODE_DECRYPT) {
+      throw new Error('mode must be either encrypt or decrypt');
+    }
+  }
+
+  async transformPrevChunk(isLast, controller) {
+    if (this.mode === MODE_ENCRYPT) {
+      controller.enqueue(
+        await this.encryptRecord(this.prevChunk, this.seq, isLast)
+      );
+      this.seq++;
+    } else {
+      if (this.seq === 0) {
+        //the first chunk during decryption contains only the header
+        const header = this.readHeader(this.prevChunk);
+        this.salt = header.salt;
+        this.rs = header.rs;
+        this.key = await this.generateKey();
+        this.nonceBase = await this.generateNonceBase();
+      } else {
+        controller.enqueue(
+          await this.decryptRecord(this.prevChunk, this.seq - 1, isLast)
+        );
+      }
+      this.seq++;
+    }
+  }
+
+  async transform(chunk, controller) {
+    if (!this.firstchunk) {
+      await this.transformPrevChunk(false, controller);
+    }
+    this.firstchunk = false;
+    this.prevChunk = Buffer.from(chunk.buffer);
+  }
+
+  async flush(controller) {
+    if (this.prevChunk) {
+      await this.transformPrevChunk(true, controller);
+    }
+  }
+}
+
+class BlobSlicer {
+  constructor(blob, rs, mode) {
+    this.blob = blob;
+    this.index = 0;
+    this.mode = mode;
+    this.chunkSize = mode === MODE_ENCRYPT ? rs - 17 : rs;
+  }
+
+  pull(controller) {
+    return new Promise((resolve, reject) => {
+      const bytesLeft = this.blob.size - this.index;
+      if (bytesLeft <= 0) {
+        controller.close();
+        return resolve();
+      }
+      let size = 1;
+      if (this.mode === MODE_DECRYPT && this.index === 0) {
+        size = Math.min(21, bytesLeft);
+      } else {
+        size = Math.min(this.chunkSize, bytesLeft);
+      }
+      const blob = this.blob.slice(this.index, this.index + size);
+      const reader = new FileReader();
+      reader.onload = () => {
+        controller.enqueue(new Uint8Array(reader.result));
+        resolve();
+      };
+      reader.onerror = reject;
+      reader.readAsArrayBuffer(blob);
+      this.index += size;
+    });
+  }
+}
+
+class BlobSliceStream extends ReadableStream {
+  constructor(blob, size, mode) {
+    super(new BlobSlicer(blob, size, mode));
+  }
+}
+
+/*
+input: a blob containing data to be transformed
+key:  Uint8Array containing key of size KEY_LENGTH 
+mode: string, either 'encrypt' or 'decrypt'
+rs:   int containing record size, optional
+salt: ArrayBuffer containing salt of KEY_LENGTH length, optional
+*/
+export default class ECE {
+  constructor(input, key, mode, rs, salt) {
+    if (rs === undefined) {
+      rs = RS;
+    }
+    if (salt === undefined) {
+      salt = generateSalt(KEY_LENGTH);
+    }
+
+    this.streamInfo = {
+      recordSize: rs,
+      fileSize: input.size + 16 * Math.floor(input.size / (rs - 17))
+    };
+    input = new BlobSliceStream(input, rs, mode);
```

Comment:
It's usually easier to understand code when parameters aren't reassigned. Something like `const blobStream = ...` is more clear.